### ISSUE TITLE
A11y: Continuous text hyperlinks are marked by colour only and therefore not accessible

### ISSUE THREAD

**Issue by daniellekirkwood** [element-hq/element-web#24932]
Continuous text links (examples marked red) are only marked by a colored highlight.
There is no additional symbol, underlining or boldface. The required contrast ratio of at least 3:1 to the rest of the text is not met with a value of 1.7:1.

Since the perception of color is required to identify the continuous text links, they are
difficult or impossible to recognise for visually impaired users if the contrast is too
weak. This is also present elsewhere.

The contrast ratios in the alternative displays "dark" and "high contrast" also do not meet the minimum requirements.

<img width="272" alt="Screenshot 2023-03-24 at 13 15 36" src="https://user-images.githubusercontent.com/89144281/227531088-7aaafcb2-ddc8-423e-bf4f-19106cb4c17e.png">

**Comment by daniellekirkwood** [comment]
@janogarcia similarly, we need your guidance here on how we plan on changing and addressing hyperlinks in the future as a part of compound. Assigning you.

**Comment by janogarcia** [comment]
@germain-gg @daniellekirkwood 

I'll be referring to contrast values for Compound, as applied in the Colour pass Germain recently implemented:

- Blue links in messages: [Accessible](https://contrast-triangle.com/?linkColor=0467dd&textColor=%231b1d22) (3.2:1). Don't need an underline. 
- Green links elsewhere: [Accessible](https://contrast-triangle.com/?linkColor=%23007a61&textColor=%231b1d22) (3.18:1). Don't need an underline. 

So, consider it solved with Compound.

**Comment by janogarcia** [comment]
Should I close the issue already or wait for the Colour pass to be released?

**Comment by daniellekirkwood** [comment]
@janogarcia  thanks for the response but the issue here is that the links are only seen to be links denoted by colour which is not accessible. It's not the colour itself that's the problem

**Comment by janogarcia** [comment]
The links I provided above demonstrate why we don't need to show an underline by default. That is, the contrast against surrounding text is > 3:1 when using Compound colors.

On top of that, only thing that needs to happen is showing an underline on hover or alternative decoration on focus, but the color part for the links when in the default state is already sorted out by Compound colors.

**Comment by daniellekirkwood** [comment]
in that case this can be closed as the compound colours handle it. You agree, @janogarcia ?

**Comment by janogarcia** [comment]
@daniellekirkwood The concern regarding 3:1 contrast against surrounding text is definitely handled by Compound. That said, we may have some instances in the legacy codebase where those links don't get an underline when hovered (e.g., links in supporting text in `Settings > Help & About`).

Maybe we want to keep this issue open to track those improvements to the legacy app? /cc @germain-gg

**Comment by daniellekirkwood** [comment]
<img width="358" alt="Screenshot 2023-08-07 at 12 01 25" src="https://github.com/vector-im/element-web/assets/89144281/d9ae743b-514b-488b-ac51-03b02527c66c">

Here's our first example. Hovering over this does not show an underline but as to your colour comments @janogarcia  this _is_ the new colours and I didn't spot that it was a link for a while. I think that we need to be really careful here especially as we can't rely on someone using their mouse to hover over random text hoping that it'll underline.

**Comment by daniellekirkwood** [comment]
<img width="368" alt="Screenshot 2023-08-07 at 12 03 47" src="https://github.com/vector-im/element-web/assets/89144281/b7d2dc0c-9a60-49b7-b7a1-eeab30e9396d">

the links in the labs and help&about section are blue and much much clearer

**Comment by daniellekirkwood** [comment]
<img width="464" alt="Screenshot 2023-08-07 at 12 06 40" src="https://github.com/vector-im/element-web/assets/89144281/a245955b-0e6c-43d3-8e94-e7785aeba6eb">

**Comment by daniellekirkwood** [comment]
@janogarcia How would you proceed from here? Please update :)

**Comment by daniellekirkwood** [comment]
@janogarcia polite reminder...

**Comment by janogarcia** [comment]
@daniellekirkwood Apologies, I got your previous messages while I was away (for 3 weeks) and it seems I didn't notice them when I caught up with stuff.

Yeah, thanks for those examples. 

So, the [safe contrast ratios I shared](https://github.com/vector-im/element-web/issues/24932#issuecomment-1644203002) above are true when non-underlined links are combined with the primary text color, but they will fail when combined with the lower contrast secondary text color.

- [Blue + Secondary Text](https://contrast-triangle.com/?linkColor=0467dd&textColor=%23656d77)
- [Green + Secondary Text](https://contrast-triangle.com/?linkColor=%23007a61&textColor=%23656d77)

For those cases we should underline the links by default.

**Comment by janogarcia** [comment]
Alternatively, we could show underlines by default only in the high-contrast theme. That's still a perfectly valid way to handle accessibility issues/enhancements. /cc @nadonomy 

https://www.w3.org/WAI/WCAG21/quickref/?showtechniques=143#contrast-minimum

For example, Twitter will underline links when enabling the high-contrast mode on iOS system-wide ("Increase Contrast" setting in iOS preferences) or switching on the "Increase color contrast" preference on Web.

<img width="633" alt="SCR-20230913-jyxh" src="https://github.com/vector-im/element-web/assets/60790/1b0e5ac4-5c02-4f3b-a1a5-e485b529a752">

**Comment by janogarcia** [comment]
Also, the [Security Disclosure Policy example](https://github.com/vector-im/element-web/issues/24932#issuecomment-1667649436) you shared would still be valid/accessible, as it's using a icon to indicate the link and not relying exclusively on color.

**Comment by janogarcia** [comment]
Update on the high-contrast theme.

We basically need to underline the links by default on the high-contrast theme. Otherwise the contrast for links against surrounding text will be low even for the primary text color.

https://contrast-triangle.com/?linkColor=%230543a7&textColor=%231a1c21&underlines=false

**Comment by daniellekirkwood** [comment]
Thanks for your comments. Is that the only action that's needed here? To add an underline when in high-contrast mode? 
Are we comfortable having the difference between modes?

**Comment by daniellekirkwood** [comment]
<img width="954" alt="Screenshot 2023-09-19 at 16 53 45" src="https://github.com/vector-im/element-web/assets/89144281/a911cd61-0bdb-4bca-8996-5529d91798e4">

FYI @janogarcia  this report shows that we should have the underline everywhere...

**Comment by daniellekirkwood** [comment]
@janogarcia polite reminder that we're waiting for your response to move this forward.

**Comment by janogarcia** [comment]
Thanks for pinging, @daniellekirkwood.

I'm not sure if the report is accurate in that detail. Links in running text can be accessible, even if the only difference is in color, as long as that the relative contrast between links and surrounding text is 3:1.

That said, for the case of links in running text:

- on desktop, we still need an underline (or alternative accessible indication) when hovering the link;
- it fails when the surrounding text is set to the secondary text color (gray instead of almost black); and,
- it fails when in the high-contrast theme, since the link color will also be darker and thus won't meet the minimum 3:1 contrast against surrounding text

Given there are so many scenarios where non-underlined links in running text can fail to meet the accessibility guidelines, I think we basically have two possible directions:

- **Option A** — Providing a setting to enable link underlines (e.g., GitHub), or just enabling it automatically for the high-contrast theme (ala Twitter).
- **Option B** — Just render underlines by default for links in running text.

→ I'd say Option B, given that Option A makes the default experience less accessible (even if it's still compliant by providing a setting).

Standalone links (e.g., menus, footers... see footer elements in GitHub screenshot) are not affected by this rule, you just need the text to be 4.5:1 against the background.

---

Refer to the currently introduced setting on GitHub and their [accessibility guidelines for links](https://primer.style/guides/accessibility/links).

They use a CSS selector (`link--inTextBlock`) for targeting links in running text.

<img width="364" alt="SCR-20231019-nqsp" src="https://github.com/vector-im/element-web/assets/60790/27e046fd-cde4-4842-a465-a704d78817fc">

<img width="632" alt="SCR-20231019-nrim" src="https://github.com/vector-im/element-web/assets/60790/2ea76d3c-d91c-4c8e-b0c9-586c3813d3f2">

<img width="1218" alt="SCR-20231124-lvec" src="https://github.com/vector-im/element-web/assets/60790/e26633a4-ca2b-4d46-9e63-af481a83f4b2">

**Comment by janogarcia** [comment]
> You're now advising that we underline all hyperlinks on web?

In running text we should always render the underline. Example on GitHub:

<img width="743" alt="link-embedded" src="https://github.com/vector-im/element-web/assets/60790/4eaa141a-6d3f-401d-b562-e1c6d8320a41">

Standalone links are not required to be underlined. From accessibility point of view, these are optional. We render an underline for most of them for the new UIs we're working on, but we may also omit it for low priority elements that should have lower visual weight (e.g. page footers). Example on GitHub:

<img width="524" alt="link-standalone" src="https://github.com/vector-im/element-web/assets/60790/ef3da5b9-fa05-41e5-8341-2783e68716c0">

> And should we be going for blue or green?

At the design-system level we use mostly black, but use blue for external or user-generated links (e.g., links in user-generated content in the timeline). Similar example on GitHub:

<img width="306" alt="SCR-20231128-pfkd" src="https://github.com/vector-im/element-web/assets/60790/91a0b738-5d0e-4d47-be7e-e7c49aaa304f">

**Comment by janogarcia** [comment]
Additional note on the colors used for links in the design system vs. legacy Element Web.

We don't need to hunt down every link in the legacy Element Web app and map it to either black or blue. For now it's OK to just make sure we're showing the underline where needed.

So, in the following example we'd just make sure the green link shows an underline, but we don't need to change it to black or blue just yet.

As we redesign the different areas of Element Web we'll make sure to update those elements as needed. For now consider it being a first pass where we just make sure we show the underline where needed.

**Comment by daniellekirkwood** [comment]
I need to confirm with the customer whether Jano's statement above is true or not. Assigning to me to do that.

@gaelledel assigning to you because: **Can you confirm that the compound rules and guidance has been updated in all the necessary places to ensure that hyperlinks everywhere are shown appropriately given the above discussion?**

**Comment by gaelledel** [comment]
Yes, the [compound design system ](https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=641%3A3272&mode=design&t=zbzVQAHDs5016qQu-1) is using underlines under text based  components
<img width="480" alt="Screenshot 2023-12-06 at 15 08 26" src="https://github.com/vector-im/element-web/assets/76945183/540fc65d-f0b2-4bee-ba5f-96f68827a3a6">

**Comment by nadonomy** [comment]
> I need to confirm with the customer whether Jano's statement above is true or not. Assigning to me to do that.
> 
> @gaelledel assigning to you because: **Can you confirm that the compound rules and guidance has been updated in all the necessary places to ensure that hyperlinks everywhere are shown appropriately given the above discussion?**

Added this issue to track adding a blue external text link issue variant in Compound: https://github.com/vector-im/compound/issues/286

Not clear when we pick this up though (design, engineering) - probably best for us to discuss in the next web leads sync.

**Comment by daniellekirkwood** [comment]
Great. As we now have that issue to track this update, I don't think we need this one anymore so I propose we close it. Thoughts?

**Comment by gaelledel** [comment]
I think it's good to be closed
